contracts/modules/UniswapManager.sol

```diff
@@ -0,0 +1,159 @@
+pragma solidity ^0.5.4;
+import "../wallet/BaseWallet.sol";
+import "./common/BaseModule.sol";
+import "./common/RelayerModule.sol";
+import "./common/OnlyOwnerModule.sol";
+import "../exchange/ERC20.sol";
+import "../utils/SafeMath.sol";
+import "../utils/SafeMath.sol";
+import "../storage/GuardianStorage.sol";
+
+interface UniswapFactory {
+    function getExchange(address _token) external view returns(address);
+}
+
+/**
+ * @title UniswapManager
+ * @dev Contract enabling the wallet owner to pool tokens to a Uniswap liquidity pool.
+ * @author Julien Niset - <julien@argent.im>
+ */
+contract UniswapManager is BaseModule, RelayerModule, OnlyOwnerModule {
+
+    bytes32 constant NAME = "UniswapManager";
+
+    using SafeMath for uint256;
+
+    // The Uniswap factory contract
+    UniswapFactory uniswap;
+    // The Guardian storage 
+    GuardianStorage public guardianStorage;
+
+    /**
+     * @dev Throws if the wallet is locked.
+     */
+    modifier onlyWhenUnlocked(BaseWallet _wallet) {
+        require(!guardianStorage.isLocked(_wallet), "TT: wallet must be unlocked");
+        _;
+    }
+
+    constructor(
+        ModuleRegistry _registry, 
+        GuardianStorage _guardianStorage, 
+        address _uniswap
+    ) 
+        BaseModule(_registry, NAME) 
+        public 
+    {
+        guardianStorage = GuardianStorage(_guardianStorage);
+        uniswap = UniswapFactory(_uniswap);
+    }
+ 
+    /**
+     * @dev Adds liquidity to a Uniswap ETH-ERC20 pair.
+     * @param _wallet The target wallet
+     * @param _poolToken The address of the ERC20 token of the pair.
+     * @param _ethAmount The amount of ETH available.
+     * @param _tokenAmount The amount of ERC20 token available.
+     * @return the number of liquidity shares minted. 
+     */
+    function addLiquidityToUniswap(
+        BaseWallet _wallet, 
+        address _poolToken, 
+        uint256 _ethAmount, 
+        uint256 _tokenAmount
+    )
+        external 
+        onlyOwner(_wallet)
+        onlyWhenUnlocked(_wallet)
+    {
+        address pool = uniswap.getExchange(_poolToken);
+        require(pool != address(0), "UM: The target token is not traded on Uniswap");
+
+        uint256 ethPoolSize = address(pool).balance;
+        uint256 tokenPoolSize = ERC20(_poolToken).balanceOf(pool);
+
+        uint256 tokenInEth = getInputToOutputPrice(_tokenAmount, tokenPoolSize, ethPoolSize);
+        if(_ethAmount >= tokenInEth) {
+            // swap some eth for tokens
+            (uint256 ethSwap, uint256 ethPool, uint256 tokenPool) = computePooledValue(ethPoolSize, tokenPoolSize, _ethAmount, _tokenAmount, tokenInEth);
+            _wallet.invoke(pool, ethSwap, abi.encodeWithSignature("ethToTokenSwapInput(uint256,uint256)", 1, block.timestamp));
```

Comment:
shouldn't we do the trade via our kyber module instead, to offer better rates to our users?

Reply:
It's a fair point and I thought about it.

Kyber: simpler math (the swap does not affect the size of the pool) and potentially a better price (although arbitrage should always mitigate that).  

Uniswap: much cheaper in terms of gas (~40k vs 150k)... and I kind of liked the idea of doing everything with Uniswap :-).

Reply:
Ok fair enough.
Another thing: I think the call to `_wallet.invoke(pool, ethSwap, abi.encodeWithSignature("ethToTokenSwapInput(uint256,uint256)", 1, block.timestamp)); ` should be wrapped in an `if(ethSwap > 0){}` because in the (rare) cases where `ethSwap == 0`,   `ethToTokenSwapInput` will revert (see L128 in uniswap_exchange.vy).

Reply:
Agreed and done.